internal/policies/mount/adsys_mount/src/lib.rs

```diff
@@ -0,0 +1,225 @@
+        main_loop.quit();
+    }
+    glib::Continue(*mounts_left != 0)
+}
+
+/// Reads the file and parses the mount points listed in it.
+fn parse_entries(path: &str) -> Result<Vec<Entry>, std::io::Error> {
+    debug!("Parsing file {} content", path);
+
+    let mut parsed_entries: Vec<Entry> = Vec::new();
+
+    let content = fs::read_to_string(path)?;
+    for line in content.lines() {
+        if line.is_empty() {
+            continue;
+        }
+
+        parsed_entries.push(match line.strip_prefix("[krb5]") {
+            Some(s) => Entry {
+                path: s.to_string(),
+                krb_auth: true,
+            },
+            None => Entry {
+                path: line.to_string(),
+                krb_auth: false,
+            },
+        });
+    }
+
+    Ok(parsed_entries)
+}
+
+/// Handles the mount operation to mount the specified entry.
+fn mount_entry(entry: Entry, tx: glib::Sender<Msg>) {
+    debug!("Mounting entry {}", entry.path);
+
+    let f = gio::File::for_uri(&entry.path);
+
+    let mount_op = gio::MountOperation::new();
+
+    mount_op.set_anonymous(true);
+    if entry.krb_auth {
+        debug!("Kerberos authentication required for {}", entry.path);
+        mount_op.set_anonymous(false);
+    }
+
+    mount_op.connect_ask_password(|mount_op, _, _, _, flags| unsafe {
+        ask_password_cb(mount_op, "msg", "user", "domain", flags)
+    });
+
+    // Callback invoked by gio after setting up the mount.
+    let mount_handled_cb = move |r: Result<(), glib::Error>| {
+        let msg = Msg {
+            path: entry.path,
+            status: r,
+        };
+
+        if let Err(e) = tx.send(msg) {
+            error!("Failed to send message in the channel: {}", e);
+        }
+    };
+
+    f.mount_enclosing_volume(
+        gio::MountMountFlags::NONE,
+        Some(&mount_op),
+        gio::Cancellable::NONE,
+        mount_handled_cb,
+    );
+}
+
+/// Callback that is invoked by gio when prompted for password.
+unsafe fn ask_password_cb(
+    mount_op: &gio::MountOperation,
+    _: &str,
+    _: &str,
+    _: &str,
+    flags: gio::AskPasswordFlags,
+) {
+    if mount_op.is_anonymous() && flags.contains(gio::AskPasswordFlags::ANONYMOUS_SUPPORTED) {
+        // SAFETY: The usage of set_data and data can result in type unsafety. Since we are
+        // controlling which key is being set and by whom, this should be safe.
+        unsafe {
+            if let Some(data) = mount_op.data::<bool>("state") {
+                // Ensures that we only try anonymous access once.
+                if *(data.as_ref()) {
+                    warn!("Anonymous access denied.");
+                    mount_op.reply(gio::MountOperationResult::Aborted);
+                }
+            } else {
+                debug!("Anonymous is supported by the provider.");
+                mount_op.set_data("state", true);
+                mount_op.reply(gio::MountOperationResult::Handled);
+            }
+        }
```

Comment:
I highly recommend limiting the `unsafe` scope so that only the statements requiring `unsafe` are inside the `unsafe` block, making it more obvious which operations may be prone to issues.

```suggestion
        // SAFETY: The usage of set_data and data can result in type unsafety. Since we are
        // controlling which key is being set and by whom, this should be safe.
        if let Some(data) = unsafe { mount_op.data::<bool>("state") } {
            // Ensures that we only try anonymous access once.
            if *(data.as_ref()) {
                warn!("Anonymous access denied.");
                mount_op.reply(gio::MountOperationResult::Aborted);
            }
        } else {
            debug!("Anonymous is supported by the provider.");
            mount_op.set_data("state", true);
            mount_op.reply(gio::MountOperationResult::Handled);
        }
```